### ISSUE TITLE
Update .w3i Info file parser

### ISSUE THREAD

**Issue by Luashine** [inwc3/wc3libs#80]
The Map info file has been fully documented now, format versions 0-31. The missing legacy pieces have literally never seen the light of day.

- PR: ChiefOfGxBxL/WC3MapSpecification/pull/7
- Until merged, [available on my fork](https://github.com/Luashine/WC3MapSpecification/blob/w3i/Info/0-31.md)

After implementing this, you'll be able to correctly read the info of ROC/TFT Beta maps.

----

This would be a total rewrite of the hardcoded [W3I.java](https://github.com/inwc3/wc3libs/blob/c86fede635cd46bca074a3280111baf44e54ce2e/src/main/java/net/moonlightflower/wc3libs/bin/app/W3I.java)

**Comment by Frotty** [comment]
Thank you for letting us know. I appreciate your suggestion, but at this time, I think that implementing this feature is not a priority for us. However, if you or anyone else would like to contribute to the codebase and make this change, we would be happy to review a pull request.

**Comment by WaterKnight** [comment]
I don't have any such old/odd formatted files for testing, though.

**Comment by WaterKnight** [comment]
> This would be a total rewrite of the hardcoded [W3I.java](https://github.com/inwc3/wc3libs/blob/c86fede635cd46bca074a3280111baf44e54ce2e/src/main/java/net/moonlightflower/wc3libs/bin/app/W3I.java)

Well, the scheme there is that every format version will have its own parsing method except that the reusable chunks re-use some code. This may be not so well suited for small changes, but I was planning to do the parsing at some point differently anyway, that the W3I would only be the model class and serializers/deserializers are extracted to better support different formats and those can be generated from specifications or make use of standard serialization/deserialization libraries.

**Comment by Luashine** [comment]
In Warcraft the parser evolved around backwards-compatibility, like around a *living format*. The parser is a single entity with some branching for version support and does set default values where needed. This is the reason Hodor wrote his documentation the way he did and I picked up where he left.

**Comment by WaterKnight** [comment]
Yes, I think it's good to write the documentation this way. If each version was separate, there would likely be drift and it's worse for interfacing. In wc3libs, I also did it this way that you can load files of different formats that map to the same model class and the fields not available in the serialized form are just null/default in the model then. Later, you might want to output it in a different format than it was originally read in, so stuff required in the serialized form missing in the model will again end up writing defaults or if some fields are not supported in the serialized form, those will just be omitted.

We can do this branching strategy as well in the parsers/writers, but when they are auto-generated from a schema that supports branching itself, it does not matter that much.
